fix: cursor movement logic inconsistencies

0Ky opened the pull request:
### Bug Fixes:
- Fixed cursor movement logic to allow movement to the next line when the right arrow key is pressed at the end of a line, and vice versa.
- Fixed an inconsistency in cursor navigation across lines of varying lengths. The cursor now retains its position based on the shorter line when moving between long and short lines, rather than reverting to its original position.
- Prevented the cursor from moving beyond the start or end of the document.

### Enhancements:
- Used the `len()` method on the `rows` vector of the `Document` struct to determine the total number of rows in the document, ensuring accurate cursor movement and boundary checking.

yhyadev reviewed (changes requested):
It's just overflows the document current column length

0Ky commented:
> It's just overflows the document current column length

I have made changes, can you please review?

- Modify cursor's column position in set_cursor to account for scroll_offset.column.
- Ensure cursor is correctly positioned on the screen during horizontal scrolls.
- Properly adjust cursor column position when transitioning between lines, this closes #1
- Address potential visibility issues due to both horizontal and vertical scrolls.

yhyadev reviewed (changes requested):
Could you store the position history? We need it to remember the past cursor position, But i don't see any history here

0Ky commented:
> Could you store the position history? We need it to remember the past cursor position, But i don't see any history here

Do you want me to create another PR for that? Your request seems like an enhancement that could be used for other features that are yet to be implemented. The current PR is focused on fixing the current "cursor movement logic inconsistencies".

yhyadev commented:
> > Could you store the position history? We need it to remember the past cursor position, But i don't see any history here
> 
> Do you want me to create another PR for that? Your request seems like an enhancement that could be used for other features that are yet to be implemented. The current PR is focused on fixing the current "cursor movement logic inconsistencies".

It was a feature before you made this PR so it's not a new thing to make a new PR for it @0Ky

yhyadev reviewed (changes requested):
This doesn't change anything, Just a new variable

yhyadev reviewed (changes requested):
Also that's just a variable

0Ky commented:
> This doesn't change anything, Just a new variable

I was under the impression that you were referring to that variable, can you show the exact commit or line that you're talking about? What exactly was the "feature" you had before I made the PR which you believe removed it?

yhyadev commented:
> > This doesn't change anything, Just a new variable
> 
> I was under the impression that you were referring to that variable, can you show the exact commit or line that you're talking about? What exactly was the "feature" you had before I made the PR which you believe removed it?

While you are going to right
```rs
self.editor.position.column = self.editor.position.column.saturating_add(1);
                    self.editor.position.history.column =
                        self.editor.position.history.column.saturating_add(1);
```

When you go down
```rs
                    self.editor.position.row = self.editor.position.row.saturating_add(1);
                    self.editor.position.column =
                        std::cmp::min(self.editor.position.history.column.saturating_sub(self.editor.scroll_offset.column), next_row_length.saturating_sub(self.editor.scroll_offset.column));
                }
```

It will choose between the minimum value between the current document length and the position history

0Ky commented:
> It will choose the minimum value between the current document length and the position history of the previous column

Oh, when you move up or down between lines of different lengths, you want to remember the horizontal position they were last on (even if the new line is shorter), so that if they return to a longer line, the cursor will attempt to go back to its remembered horizontal position.

yhyadev reviewed (changes requested):
Ok, We done with the fancy thing, Let's begin to really scroll up and down, Because we are now able to scroll left and right only

0Ky commented:
@yhyadev Oke, Here are the changes that were made:

- Added support for scrolling up and down for all the keys.
- Made further refinement of the cursor movement logic.
- Fixed a conditional logic error that prevented drawing the first line of the document.
- Removed unused variables.

yhyadev reviewed (approved):
LGTM 👍
